hashing/ripemd_160.cpp

```diff
@@ -0,0 +1,437 @@
+
+        int32_t A = 0, B = 0, C = 0, D = 0, E = 0, T = 0;
+        int32_t A_dash = 0, B_dash = 0, C_dash = 0, D_dash = 0, E_dash = 0,
+                T_dash = 0;
+
+        // for each 512 bits of data
+        for (uint64_t i = 0; i < word_data.size(); i += 16) {
+            current_block_data = std::vector<uint32_t>(
+                word_data.begin() + i, word_data.begin() + i + 16);
+
+            A = h_0;
+            B = h_1;
+            C = h_2;
+            D = h_3;
+            E = h_4;
+
+            A_dash = h_0;
+            B_dash = h_1;
+            C_dash = h_2;
+            D_dash = h_3;
+            E_dash = h_4;
+
+            // parallely process the 80 rounds
+            std::thread compute_for_block_thread(&compute_for_block, this, &A,
+                                                 &B, &C, &D, &E, &T,
+                                                 std::ref(current_block_data));
+            std::thread compute_for_block_dash_thread(
+                &compute_for_block_dash, this, &A_dash, &B_dash, &C_dash,
+                &D_dash, &E_dash, &T_dash, std::ref(current_block_data));
+
+            compute_for_block_thread.join();
+            compute_for_block_dash_thread.join();
+
+            // update the current state after 80 rounds
+            T = h_1 + C + D_dash;
+            h_1 = h_2 + D + E_dash;
+            h_2 = h_3 + E + A_dash;
+            h_3 = h_4 + A + B_dash;
+            h_4 = h_0 + B + C_dash;
+            h_0 = T;
+        }
+
+        // convert the bytes in little endian format to hex in string format
+        std::string hash = "";
+        hash = hash + uint32_t_to_string(h_0);
+        hash = hash + uint32_t_to_string(h_1);
+        hash = hash + uint32_t_to_string(h_2);
+        hash = hash + uint32_t_to_string(h_3);
+        hash = hash + uint32_t_to_string(h_4);
+
+        return hash;
+    }
+};
+}  // namespace hashing
+
+/**
+ * @brief self test implementation
+ * @return void
+ */
+static void test() {
+    hashing::RIPEMD160 obj;
+    assert(obj.hash("").compare("9c1185a5c5e9fc54612808977ee8f548b2258d31") ==
+           0);
+    assert(obj.hash("a").compare("0bdc9d2d256b3ee9daae347be6f4dc835a467ffe") ==
+           0);
+    assert(obj.hash("abc").compare(
+               "8eb208f7e05d987a9b044a8e98c6b087f15a0bfc") == 0);
+    assert(obj.hash("message digest")
+               .compare("5d0689ef49d2fae572b881b123a85ffa21595f36") == 0);
+    assert(obj.hash("abcdefghijklmnopqrstuvwxyz")
+               .compare("f71c27109c692c1b56bbdceb5b9d2865b3708dbc") == 0);
+    assert(obj.hash("abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq")
+               .compare("12a053384a9c0c88e405a06c27dcf49ada62eb2b") == 0);
+    assert(
+        obj.hash(
+               "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789")
+            .compare("b0e20b6e3116640286ed3a87a5713079b21f5189") == 0);
+    assert(obj.hash("1234567890123456789012345678901234567890123456789012345678"
```

Comment:
add tests where the error case is handled
these tests should be covered via 
```cpp
try {
 hash("error_prone_input");
} catch(const std::invalid_argument& err) {
  assert(err.what() == "j value out of bound");
}

Reply:
the errors can never occur. it is a internally invoked function. A private function. j value is always within bounds. I wrote it just to suppress the warnings. It can be proved that j value will be in bounds. Because the number of rounds in compression is 80.